internalsdk/session_model.go

```diff
@@ -725,3 +1154,735 @@ func checkAdsEnabled(session *SessionModel) error {
+	// // a malicious B sent from server
+	B := big.NewInt(0).SetBytes(srpB.B)
+
+	if err = client.SetOthersPublic(B); err != nil {
+		log.Errorf("Error while setting srpB %v", err)
+		return err
+	}
+
+	// client can now make the session key
+	clientKey, err := client.Key()
+	if err != nil || clientKey == nil {
+		return log.Errorf("user_not_found error while generating Client key %v", err)
+	}
+
+	// // check if the server proof is valid
+	if !client.GoodServerProof(salt, lowerCaseEmail, srpB.Proof) {
+		return log.Errorf("user_not_found error while checking server proof%v", err)
+	}
+
+	clientProof, err := client.ClientProof()
+	if err != nil {
+		return log.Errorf("user_not_found error while generating client proof %v", err)
+	}
+
+	changeEmailRequestBody := &protos.ChangeEmailRequest{
+		OldEmail: lowerCaseEmail,
+		NewEmail: lowerCaseNewEmail,
+		Proof:    clientProof,
+	}
+
+	isEmailChanged, err := session.proClient.ChangeEmail(context.Background(), changeEmailRequestBody)
+	if err != nil {
+		return err
+	}
+	log.Debugf("Change Email response %v", isEmailChanged)
+	return nil
+}
+
+func completeChangeEmail(session SessionModel, email string, newEmail string, password string, code string) error {
+	// Create new salt and verifier with new email and new slat
+	newsalt, err := GenerateSalt()
+	if err != nil {
+		return err
+	}
+	log.Debugf("Slat %v and length %v", newsalt, len(newsalt))
+
+	srpClient := srp.NewSRPClient(srp.KnownGroups[group], GenerateEncryptedKey(password, newEmail, newsalt), nil)
+	verifierKey, err := srpClient.Verifier()
+	if err != nil {
+		return err
+	}
+
+	completeChangeEmail := &protos.CompleteChangeEmailRequest{
+		OldEmail:    email,
+		NewEmail:    newEmail,
+		NewSalt:     newsalt,
+		NewVerifier: verifierKey.Bytes(),
+		Code:        code,
+	}
+
+	isEmailChanged, err := session.authClient.CompleteChangeEmail(context.Background(), completeChangeEmail)
+	if err != nil {
+		return err
+	}
+	log.Debugf("Compelte Change Email response %v", isEmailChanged)
+	return setEmail(session.baseModel, newEmail)
+}
+
+// Clear slat and change accoutn state
+func signOut(session SessionModel) error {
+	email, err := session.Email()
+	if err != nil {
+		return log.Errorf("Email not found %v", err)
+	}
+
+	deviceId, err := session.GetDeviceID()
+	if err != nil {
+		return log.Errorf("deviceId not found %v", err)
+	}
+
+	token, err := session.GetToken()
+	if err != nil {
+		return log.Errorf("token not found %v", err)
+	}
+
+	userId, err := session.GetUserID()
+	if err != nil {
+		return log.Errorf("userid not found %v", err)
+	}
+
+	signoutData := &protos.LogoutRequest{
+		Email:        email,
+		DeviceId:     deviceId,
+		LegacyToken:  token,
+		LegacyUserID: userId,
+	}
+
+	log.Debugf("Sign out request %+v", signoutData)
+
+	loggedOut, logoutErr := session.authClient.SignOut(context.Background(), signoutData)
+	if err != nil {
```

Comment:
I think you want to check if logoutErr is nil here instead of err?

Reply:
Thanks for caching this, Done.